### PR TITLE
SUPESC-781 Fixed duplicated preauthorisation calls.

### PR DESCRIPTION
**Pull request by olhalivitchuk**
- Developer(s): @olhalivitchuk 

- Ticket: https://spryker.atlassian.net/browse/SUPESC-781

- Release Group: https://release.spryker.com/release-groups/view/5351

- PR Overview: https://release.spryker.com/release/pull-request-eco/Payone/88

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Payone                | minor                 |                      |

-----------------------------------------

##### Change log

Improvements

- Increased minimal required `PHP` version to 8.1.

Adjustments

- Adjusted `PayoneFacadeInterface::orderPostSave()` so it verifies that a preauthorization request has been already sent instead of sending it twice.

### CODE REVIEW

File: src/SprykerEco/Zed/Payone/Business/Payment/RequestSender/PayoneBaseAuthorizeSender.php [hunk 1]
```diff
@@ -68,7 +77,12 @@ public function performAuthorizationRequest(
         $this->standartParameterMapper->setStandardParameter($requestContainer, $this->standardParameter);
 
         $apiLogEntity = $this->initializeApiLog($paymentEntity, $requestContainer);
-        $rawResponse = $this->executionAdapter->sendRequest($requestContainer);
+        $rawResponse = $this->repository->getPreauthorizedPaymentByReference($requestContainer->getReference());
```
Comment (vol4onok) [inline src/SprykerEco/Zed/Payone/Business/Payment/RequestSender/PayoneBaseAuthorizeSender.php:80]: Sender shouldn't check anything. pls move it `$this->repository->getPreauthorizedPaymentByReference($requestContainer->getReference());` to one level app before calling `performAuthorizationRequest`

File: src/SprykerEco/Zed/Payone/Dependency/Service/PayoneToUtilEncodingServiceBridge.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+     * @param string $jsonValue
+     * @param bool $assoc Deprecated: `false` is deprecated, always use `true` for array return.
+     * @param int|null $depth
+     * @param int|null $options
+     *
+     * @return object|array<mixed>|null
+     */
+    public function decodeJson(string $jsonValue, bool $assoc = false, ?int $depth = null, ?int $options = null)
+    {
+        if ($assoc === false) {
```
Comment (vol4onok) [inline src/SprykerEco/Zed/Payone/Dependency/Service/PayoneToUtilEncodingServiceBridge.php:37]: Why do we need it here? remove it pls.

File: src/SprykerEco/Zed/Payone/Persistence/PayoneRepository.php [hunk 3]
```diff
@@ -168,6 +175,36 @@ public function createPaymentPayoneQueryByOrderId(int $idOrder): SpyPaymentPayon
...
+            ->withColumn(SpyPaymentPayoneApiLogTableMap::COL_RAW_RESPONSE, static::RAW_RESPONSE_COLUMN_NAME);
+        $paymentData = $paymentPayoneQuery->findOne();
+
+        if ($paymentData === null) {
+            return [];
+        }
+
+        $formattedPaymentData = array_merge(
+            $paymentData->toArray(),
+            $this->getFactory()->getServiceUtilEncoding()->decodeJson($paymentData->getVirtualColumn(static::RAW_RESPONSE_COLUMN_NAME), true),
```
Comment (vol4onok) [inline src/SprykerEco/Zed/Payone/Persistence/PayoneRepository.php:201]: check the result of `decodeJson` it can be null

**Review by vol4onok — COMMENTED** [review]
Do we need test here?

**Review by vol4onok — COMMENTED** [review]
comments

File: src/SprykerEco/Zed/Payone/Dependency/Service/PayoneToUtilEncodingServiceBridge.php (hunk 2 above)
Comment (yaroslav-spryker) [reply src/SprykerEco/Zed/Payone/Dependency/Service/PayoneToUtilEncodingServiceBridge.php:37]: We need it in order our old bridges to support/force using newer versions according to our arch sniffer

**Review by yaroslav-spryker — APPROVED** [review]

File: src/SprykerEco/Zed/Payone/Persistence/PayoneRepository.php [hunk 4]
```diff
@@ -168,6 +175,30 @@ public function createPaymentPayoneQueryByOrderId(int $idOrder): SpyPaymentPayon
...
+                ->filterByStatus(PayoneApiConstants::RESPONSE_TYPE_APPROVED)
+            ->endUse()
+            ->withColumn(SpyPaymentPayoneApiLogTableMap::COL_RAW_RESPONSE, static::RAW_RESPONSE_COLUMN_NAME);
+        $paymentEntity = $paymentPayoneQuery->findOne();
+
+        if ($paymentEntity === null) {
+            return [];
+        }
+
+        return $this->getFormattedPaymentData($paymentEntity->toArray());
```
Comment (vol4onok) [inline src/SprykerEco/Zed/Payone/Persistence/PayoneRepository.php:199]: why don't you use entity?

**Review by vol4onok — APPROVED** [review]

File: src/SprykerEco/Zed/Payone/Persistence/PayonePersistenceFactory.php [hunk 5]
```diff
@@ -79,4 +81,12 @@ public function createPayonePersistenceMapper(): PayonePersistenceMapper
     {
         return new PayonePersistenceMapper();
     }
+
+    /**
+     * @return \SprykerEco\Zed\Payone\Dependency\Service\PayoneToUtilEncodingServiceInterface
+     */
+    public function getServiceUtilEncoding(): PayoneToUtilEncodingServiceInterface
```
Comment (asmarovydlo) [inline src/SprykerEco/Zed/Payone/Persistence/PayonePersistenceFactory.php:88]: `Service` word should be in the beginning only for the constant.
```suggestion
    public function getUtilEncodingService(): PayoneToUtilEncodingServiceInterface
```

File: src/SprykerEco/Zed/Payone/Persistence/PayoneRepository.php (hunk 4 above)
Comment (asmarovydlo) [inline src/SprykerEco/Zed/Payone/Persistence/PayoneRepository.php:199]: Please do what @vol4onok  suggested https://github.com/spryker-eco/payone/pull/88/files#r1528958070

File: tests/SprykerEcoTest/Zed/Payone/Business/Order/HookTest.php [hunk 6]
```diff
@@ -40,4 +41,23 @@ public function testPostSaveHook()
...
+        $this->createPayonePayment();
+        $paymentApiLog = $this->createPayoneApiLog(PayoneApiConstants::REQUEST_TYPE_PREAUTHORIZATION, PayoneApiConstants::RESPONSE_TYPE_APPROVED);
+
+        $checkoutResponseTransfer = new CheckoutResponseTransfer();
+        $this->quoteTransfer->getPayment()
+            ->getPayone()
+            ->setFkSalesOrder($this->orderEntity->getIdSalesOrder());
+        $newCheckoutResponseTransfer = $this->payoneFacade->postSaveHook($this->quoteTransfer, $checkoutResponseTransfer);
+
+        $preauthorizedRequestsCount = (new SpyPaymentPayoneApiLogQuery())->filterByTransactionId('213552995')->find()->count();
```
Comment (asmarovydlo) [inline tests/SprykerEcoTest/Zed/Payone/Business/Order/HookTest.php:59]: Where does this number come from? Please make it explicit.

**Review by asmarovydlo — CHANGES_REQUESTED** [review]

**Review by asmarovydlo — APPROVED** [review]

**Pull request merged**